### PR TITLE
fredbaa bearded-shame

### PR DESCRIPTION
**Pull request by fredbaa**
- updated User model that removes unneccesary columns
- support to create/add user
- support for changing status
- support for 'massaged' feature
- added rspec, and test cases using rspec in spec folder
- fixed error on rake task and stylesheet loading
- answered questions.txt

### CODE REVIEW

File: questions.txt [hunk 1]
```diff
@@ -1,51 +1,89 @@
...
+- Optimize fetching data.
+  - Find n+1 queries when fetching data, and refactor code to have few queries as possible.
+  - If the data can be loaded via cache (memory or browser cache), then use it to your advantage.
+- Optimize assets included on page
+  - If there are assets that are not included on the page, then remove them. We don't want extra load time to load these unnecessary js/css files.
+  - Use an asset packager to minify/reduce these files (but making sure it still retains the information in it)
+  - If there are images, if possible, delay loading and load only when needed/viewed.
+- Create the illusion that the page is fast
+  - This is more on perceived faster page load. If you have lots of data to render, apply pagination that will load the data in bits.
+  - Use ajax to your advantage in CRUD data. This will minimize additional page reload, and much fun to use.
```
Comment (fredbaa) [inline questions.txt]: This is actually 3 ways, but it seems a lot. But to sum it up:
- Optimize fetching data
- Optimize assets included on page
- Create the illusion that the page is fast (using javascript to delay load everything else)

**Comment by vincentpaca** [comment]
@idoo did you falsely post that comment or something?

File: spec/models/beard_spec.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
+require 'spec_helper'
+
+describe 'Beard' do
+  before {
+    user1 = User.create(:first_name => "Paul", :last_name => "Jones")
```
Comment (vincentpaca) [inline spec/models/beard_spec.rb]: One thing you can improve here is lose the `before` and just do a: `let(:user) { User.create(:first_name => "Paul", :last_name => "Jones") }`

With this you can lose the `find_by_first_name` below.

Comment (fredbaa) [inline spec/models/beard_spec.rb]: Got it. Thanks man!

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,7 +1,9 @@
 class UsersController < ApplicationController
 
-  def list
-    @users = User.all
+  def create
+    @user  = User.create_from_params(params[:user])
+    @beard = Beard.create(:user_id => @user.id,
```
Comment (more-ron) [inline app/controllers/users_controller.rb]: Normally if you wanna be strict with object orientation you wouldn't assign by `#id`. That will become `user: @user` assignment by `#id` is the concern of the persistence layer and not you application.

File: app/models/beard.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  def update_status!(params)
+    self.status = params[:status] if params[:status]
+    self.save
+
+    self.toggle_massaged! if params[:is_massaged]
+  end
+
+  def toggle_massaged!
+    current_value = self.is_massaged
+    self.is_massaged = !!!current_value
```
Comment (more-ron) [inline app/models/beard.rb]: Overkill? One bang is enough to return `true` || `false`

File: app/models/beard.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    self.save!
+  end
+
+  def massaged_label
+    is_massaged? ? 'Has massaged' : 'Has not massaged'
+  end
```
Comment (more-ron) [inline app/models/beard.rb]: If you would want to separate out view logic out of the data model how would you do it?

File: app/models/beard.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+  private
+
+  def valid_status
+    unless STATUS.include?(self.status)
+      self.errors.add(:status_undefined, 'beard status invalid')
+    end
+  end
```
Comment (more-ron) [inline app/models/beard.rb]: Why not use below?

```ruby
validates :status, inclusion: { in: STATUS }
```

File: app/models/user.rb [hunk 7]
```diff
@@ -1,6 +1,19 @@
...
+  validates :last_name,  presence: true
+
+  def self.create_from_params(params)
+    user = self.new
+    user.first_name = params[:first_name]
+    user.last_name  = params[:last_name]
+    user.save
+
+    user
+  end
```
Comment (more-ron) [inline app/models/user.rb]: Why would you need this if you could do:

```ruby
User.create! params.permit(:first_name, :last_name)
```

File: app/models/user.rb [hunk 8]
```diff
@@ -1,6 +1,19 @@
...
 
   def fullname
-    first_name + " " + last_name
+    first_name.to_s + " " + last_name.to_s
```
Comment (more-ron) [inline app/models/user.rb]: Aren't you confident that the names would be string? Why not do interpolation?

```ruby
  "#{first_name} #{last_name}"
```

Or join that converts objects to strings before joining them.

```ruby
  [first_name, last_name].join(" ")
```

File: app/models/beard.rb (hunk 5 above)
Comment (fredbaa) [inline app/models/beard.rb]: I'd probably put it either in lib folder, create a different module that contains other 'Beard' functionalities. 

Also, this massage_label should've been on the helper, not in the model.

File: app/models/beard.rb (hunk 4 above)
Comment (fredbaa) [inline app/models/beard.rb]: That's true. This is an overkill and I don't usually do it like this. I don't know why I added extra bangs anyway. XD

Comment (more-ron) [inline app/models/beard.rb]: By the way, when do you consider if beard is massaged?

Maybe if the beard is relaxed? :trollface:

Comment (bramski) [inline app/models/beard.rb]: Maybe that's what all the bangs were for?

File: app/models/beard.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  validates :status,  presence: true
+  validates :user_id, presence: true
+
+  validate :valid_status
+
+  def update_status!(params)
+    self.status = params[:status] if params[:status]
+    self.save
+
+    self.toggle_massaged! if params[:is_massaged]
```
Comment (bramski) [inline app/models/beard.rb]: Seems like this method should be private?  Also, you don't need to use "self." for methods within the object unless there is an overscoping variable.

**Comment by vincentpaca** [comment]
Closing this PR as this has been reviewed.

**Pull request closed without merging**